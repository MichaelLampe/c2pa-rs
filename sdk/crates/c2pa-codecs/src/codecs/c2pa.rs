// Copyright 2022 Adobe. All rights reserved.
// This file is licensed to you under the Apache License,
// Version 2.0 (http://www.apache.org/licenses/LICENSE-2.0)
// or the MIT license (http://opensource.org/licenses/MIT),
// at your option.

// Unless required by applicable law or agreed to in writing,
// this software is distributed on an "AS IS" BASIS, WITHOUT
// WARRANTIES OR REPRESENTATIONS OF ANY KIND, either express or
// implied. See the LICENSE-MIT and LICENSE-APACHE files for the
// specific language governing permissions and limitations under
// each license.

use std::io::{Read, Seek, Write};

use crate::{ByteSpan, CodecError, DataHash, Decode, Encode, Hash, Span, Support};

/// Supports working with ".c2pa" files containing only manifest store data
#[derive(Debug)]
pub struct C2paCodec<R> {
    src: R,
}

impl<R> C2paCodec<R> {
    pub fn new(src: R) -> Self {
        Self { src }
    }
}

impl Support for C2paCodec<()> {
    const MAX_SIGNATURE_LEN: usize = 0;

    fn supports_signature(_signature: &[u8]) -> bool {
        false
    }

    fn supports_extension(extension: &str) -> bool {
        extension == "c2pa"
    }

    fn supports_mime(mime: &str) -> bool {
        mime == "application/c2pa" || mime == "application/x-c2pa-manifest-store"
    }
}

impl<R: Read + Seek> Decode for C2paCodec<R> {
    fn read_c2pa(&mut self) -> Result<Option<Vec<u8>>, CodecError> {
        let mut cai_data = Vec::new();
        // read the whole file
        self.src.read_to_end(&mut cai_data)?;
        Ok(Some(cai_data))
    }
}

impl<R: Read + Seek> Encode for C2paCodec<R> {
    fn write_c2pa(&mut self, mut dst: impl Write, c2pa: &[u8]) -> Result<(), CodecError> {
        // just write the store bytes and ingore the input stream
        dst.write_all(c2pa)?;
        Ok(())
    }

    fn remove_c2pa(&mut self, _dst: impl Write) -> Result<bool, CodecError> {
        // TODO: true or false?
        Ok(false)
    }

    fn patch_c2pa(&self, mut dst: impl Read + Write + Seek, c2pa: &[u8]) -> Result<(), CodecError> {
        dst.write_all(c2pa)?;
        Ok(())
    }
}

impl<R: Read + Seek> Span for C2paCodec<R> {
    fn hash(&mut self) -> Result<Hash, CodecError> {
        todo!()
    }

    fn data_hash(&mut self) -> Result<DataHash, CodecError> {
        Ok(DataHash {
            spans: vec![ByteSpan { start: 0, len: 0 }],
        })
    }
}
