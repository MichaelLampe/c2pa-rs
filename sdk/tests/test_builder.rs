// Copyright 2024 Adobe. All rights reserved.
// This file is licensed to you under the Apache License,
// Version 2.0 (http://www.apache.org/licenses/LICENSE-2.0)
// or the MIT license (http://opensource.org/licenses/MIT),
// at your option.

// Unless required by applicable law or agreed to in writing,
// this software is distributed on an "AS IS" BASIS, WITHOUT
// WARRANTIES OR REPRESENTATIONS OF ANY KIND, either express or
// implied. See the LICENSE-MIT and LICENSE-APACHE files for the
// specific language governing permissions and limitations under
// each license.

use std::{fs::File, io::Cursor};

use c2pa::{Builder, Reader, Result};

mod common;
use common::{assets::iter_assets, fixtures_path, test_signer, unescape_json};
use insta::{allow_duplicates, assert_json_snapshot};

#[test]
fn test_builder_ca_jpg() -> Result<()> {
    let manifest_def = std::fs::read_to_string(fixtures_path("simple_manifest.json"))?;
    let mut builder = Builder::from_json(&manifest_def)?;

    const TEST_IMAGE: &[u8] = include_bytes!("../tests/fixtures/CA.jpg");
    let format = "image/jpeg";
    let mut source = Cursor::new(TEST_IMAGE);

    let mut dest = Cursor::new(Vec::new());
    builder.sign(&test_signer(), format, &mut source, &mut dest)?;

    apply_filters!();
    assert_json_snapshot!(unescape_json(
        &Reader::from_stream(format, &mut dest)?.json()
    )?);

    Ok(())
}

#[test]
fn test_builder_embed_all_assets() -> Result<()> {
    // TODO: example test that runs on all sample files (seems there's an error in one of them?)
    allow_duplicates! {
        || -> Result<()> {
            for asset in iter_assets() {
                let manifest_def = std::fs::read_to_string(fixtures_path("simple_manifest.json"))?;

                let format = infer::get(asset).unwrap().mime_type();
                let mut source = Cursor::new(asset);
                let mut dest = Cursor::new(Vec::new());

                let mut builder = Builder::from_json(&manifest_def)?;
                builder.sign(&test_signer(), format, &mut source, &mut dest)?;

                apply_filters!();
                assert_json_snapshot!(unescape_json(
                    &Reader::from_stream(format, &mut dest)?.json()
                )?, {
                   ".manifests.*.format" => "[FORMAT]",
                });
            }

            Ok(())
        // TODO: https://github.com/mitsuhiko/insta/issues/530
        }().unwrap()
    }

    Ok(())
}
