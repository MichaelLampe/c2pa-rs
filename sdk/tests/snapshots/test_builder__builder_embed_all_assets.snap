---
source: sdk/tests/test_builder.rs
expression: "unescape_json(&Reader::from_stream(format, &mut dest)?.json())?"
---
{
  "active_manifest": "[MANIFEST_URN]",
  "manifests": {
    "[MANIFEST_URN]": {
      "claim_generator": "test/1.0 c2pa-rs/0.32.6",
      "claim_generator_info": [
        {
          "name": "test",
          "version": "1.0"
        },
        {
          "name": "c2pa-rs",
          "version": "0.32.6"
        }
      ],
      "format": "[FORMAT]",
      "instance_id": [XMP_ID]",
      "ingredients": [],
      "assertions": [
        {
          "label": "c2pa.actions.v2",
          "data": {
            "actions": [
              {
                "action": "c2pa.opened",
                "softwareAgent": {
                  "name": "TestApp",
                  "version": "1.0",
                  "something": "else"
                },
                "parameters": {
                  "description": "import"
                },
                "digitalSourceType": "http://cv.iptc.org/newscodes/digitalsourcetype/algorithmicMedia"
              }
            ]
          }
        }
      ],
      "signature_info": {
        "alg": "Ed25519",
        "issuer": "C2PA Test Signing Cert",
        "cert_serial_number": "638838410810235485828984295321338730070538954823"
      },
      "label": "[MANIFEST_URN]"
    }
  }
}
